dm/syncer/ddl.go

```diff
@@ -192,6 +196,53 @@ func (s *Syncer) clearOnlineDDL(tctx *tcontext.Context, targetTable *filter.Tabl
 	return nil
 }
 
+// adjustCollation adds collation for create database and check create table.
+func adjustCollation(tctx *tcontext.Context, ddlInfo *ddlInfo, statusVars []byte) {
+	// check create table has collation
+	if createStmt, ok := ddlInfo.originStmt.(*ast.CreateTableStmt); ok {
+		// create table like old table has no table_options
+		// See https://dev.mysql.com/doc/refman/5.7/en/create-table.html
+		if createStmt.ReferTable != nil {
+			return
+		}
+		for _, tableOption := range createStmt.Options {
+			// already have 'Collation'
+			if tableOption.Tp == ast.TableOptionCollate {
+				return
+			}
+		}
+		tctx.L().Warn("detect create table risk which use implicit collation", zap.String("originSQL", ddlInfo.originDDL))
+	} else if createStmt, ok := ddlInfo.originStmt.(*ast.CreateDatabaseStmt); ok {
+		// collation is in create database syntax create_option
+		// See https://dev.mysql.com/doc/refman/8.0/en/create-database.html
+		hasCharset := false
+		for _, createOption := range createStmt.Options {
+			// already have 'Collation'
+			if createOption.Tp == ast.DatabaseOptionCollate {
+				return
+			}
+			if createOption.Tp == ast.DatabaseOptionCharset {
+				hasCharset = true
+			}
+		}
+
+		// just has charset, can not add collation by server collation
```

Comment:
we should also add explicit collation because the default collation of a charset in TiDB is not same with MySQL

Reply:
ok, it will be add by `SHOW CHARACTER SET`.